tensorflow/lite/micro/kernels/unidirectional_sequence_lstm_test_config.cc

```diff
@@ -231,7 +231,7 @@ LstmIntegerTestConfig lstm_integer_no_peephole_config = {
                 kLstmSeqLengthIntegerNoPeephole *
                 kLstmNumOutputIntegerNoPeephole]){
             127, 127, -108, -67, 127, 127, -128, 127, 127, -128, 127, 127, 127,
-            127, 127, -128, 127, 127},
+            127, 13, -128, 127, 127},
```

Comment:
this is quite a large difference in the overall output compared to the previous implementation. Can you confirm that this is expected and not a bug?

Reply:
Yes. This difference is expected. Activation functions (sigmoid and tanh) are called in 5 places in LSTM. 
Input to Activations functions are constant precision(Q12 format) for first 4 calls. So, here for all potential inputs, LUT based activation functions will produce the outputs with greater accuracy. But last activation function (tanh) is called for the calculation for the cell state vector tensor. Here input precision for the  tanh function will be depend on the scale factor of cell state vector tensor. 

For the unit test case, input to this tanh function is Q15 format. Inputs needs to be down shifted to Q12  before calling LUT based tanh function since LUT based activation runs only for the Q12 input. But original gemmlowp implementation is called  directly on Q15 inputs. Although the inputs are downshifted, the LUT based activations produces higher accuracy in majority of outputs. Maximum absolute difference for cell state for the original implementation (gemmlowp)and LUT based implementation is just 6 for the first iteration. But this difference is continues to grow over the following iterations and the final difference is quite large for this unit test.
 
Below are input and outputs for the last activation function (tanh for the cell state vector calculation)  for various implementations.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/lukmanr/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/lukmanr/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{border:.5pt solid windowtext;}
.xl66
	{color:white;
	border:.5pt solid windowtext;
	background:#4472C4;
	mso-pattern:black none;}
.xl67
	{color:white;
	border:.5pt solid windowtext;
	background:#4472C4;
	mso-pattern:black none;
	white-space:normal;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



tanh   Input | Reference   calculated from actual float tanh implementation | gemmlowp | LUT based   reference_integer_ops | gemmlowp (abs   diff w.r.t ref) | LUT based   reference_integer_ops (abs diff w.r.t ref)
-- | -- | -- | -- | -- | --
-14654 | -13749 | -13744 | -13750 | 5 | 1
8937 | 8721 | 8720 | 8721 | 1 | 0
5772 | 5713 | 5708 | 5709 | 5 | 4
6717 | 6624 | 6620 | 6620 | 4 | 4
-11848 | -11357 | -11356 | -11357 | 1 | 0
7851 | 7704 | 7700 | 7701 | 4 | 3
1512 | 1510 | 1512 | 1511 | 2 | 1
12938 | 12305 | 12304 | 12304 | 1 | 1



</body>

</html>